arroyo-df/src/rewriters.rs

```diff
@@ -0,0 +1,422 @@
+        };
+
+        let table_name = table_scan.table_name.table();
+        let table = self
+            .schema_provider
+            .get_table(table_name)
+            .ok_or_else(|| DataFusionError::Plan(format!("Table {} not found", table_name)))?;
+
+        let Table::ConnectorTable(table) = table else {
+            return Ok(node);
+        };
+        let watermark_node = WatermarkNode::new(
+            self.projection(&table_scan, table)?,
+            Some(Self::watermark_expression(table)?),
+        )
+        .map_err(|err| {
+            DataFusionError::Internal(format!("failed to create watermark expression: {}", err))
+        })?;
+
+        return Ok(LogicalPlan::Extension(Extension {
+            node: Arc::new(watermark_node),
+        }));
+    }
+}
+
+pub const UNNESTED_COL: &str = "__unnested";
+
+pub struct UnnestRewriter {}
+
+impl UnnestRewriter {
+    fn split_unnest(expr: Expr) -> DFResult<(Expr, Option<Expr>)> {
+        let mut c: Option<Expr> = None;
+
+        let expr = expr.transform_up_mut(&mut |e| {
+            match &e {
+                Expr::ScalarFunction(ScalarFunction {
+                    func_def: ScalarFunctionDefinition::UDF(udf),
+                    args,
+                }) => {
+                    if udf.name() == "unnest" {
+                        match args.len() {
+                            1 => {
+                                if c.replace(args[0].clone()).is_some() {
+                                    return Err(DataFusionError::Plan(
+                                        "Multiple unnests in expression, which is not allowed"
+                                            .to_string(),
+                                    ));
+                                };
+
+                                return Ok(Transformed::Yes(Expr::Column(
+                                    Column::new_unqualified(UNNESTED_COL),
+                                )));
+                            }
+                            n => {
+                                panic!(
+                                    "Unnest has wrong number of arguments (expected 1, found {})",
+                                    n
+                                );
+                            }
+                        }
+                    }
+                }
+                _ => {}
+            };
+            Ok(Transformed::No(e))
+        })?;
+
+        Ok((expr, c))
+    }
+}
+
+impl TreeNodeRewriter for UnnestRewriter {
+    type N = LogicalPlan;
+
+    fn mutate(&mut self, node: Self::N) -> DFResult<Self::N> {
+        let LogicalPlan::Projection(projection) = &node else {
+            if node.expressions().iter().any(|e| {
+                let e = Self::split_unnest(e.clone());
+                e.is_err() || e.unwrap().1.is_some()
+            }) {
+                return Err(DataFusionError::Plan(
+                    "unnest is only supported in SELECT statements".to_string(),
```

Comment:
Maybe print something that indicates where it was found instead? Possibly just the debug for the logical plan?